SCP/Exec improvements

kontsevoy opened the pull request:
## List of changes
- `scp` implementation reuses exec infrastructure: it launches teleport via exec under the requested OS login/mapping.
- `exec` implementation separates stdout and stderr
- `exec` implementation correctly passes the process exit code and `tsh` re-uses that code in turn passing it to the caller
- error reporting has been improved: you get proper stderr without duplicates
- logs are more useful (important errors are now logged, and excessive loggign around execution has been trimmed).
## Misc bugfixes
- Invalid JSON in /var/lib/teleport/authservers.json does not prevent teleport from starting 
- That JSON file was parsed twice, I removed the dupe.
## TODO

Audit log does not provide the detailed (per-file) information about `scp` operations. You only get to see higher level scp commands.

kontsevoy committed:
Intermediate commit

kontsevoy committed:
Minor changes

kontsevoy committed:
Clean error reporting for SSH exec

- stdout and stderr are separated
- exit status is inherited by tsh

kontsevoy commented:
This PR is not ready for merging yet.

klizhentas commented on lib/client/api.go:
```diff
@@ -99,8 +98,12 @@ type Config struct {
+	Stderr io.Writer
+
+	// ExitStatus carries the returned value (exit status) of the remote
+	// process execution (via SSh exec)
```
`SSH`

klizhentas commented on lib/service/service.go:
```diff
@@ -133,18 +133,10 @@ func (process *TeleportProcess) connectToAuthService(role teleport.Role) (*Conne
 	authClient, err := auth.NewTunClient(
 		string(role),
-		authServers,
+		process.Config.AuthServers,
```
this change will defeat the purpose of the `authServers` file as every start will use the value from configuration, and as real auth servers drift away, after the restart node won't find the the servers any more.

klizhentas replied:
as a compromise you could do a merge here, making sure old configs are honored, but failover will get to new ones

kontsevoy replied:
mm... you misunderstood the purpose of this. hard to see by looking at the diff. `NewTunClient` reads "dynamic" auth servers and merges them with the passed value. this change does not affect the behaviour, it removes duplicate reading of auth servers.

klizhentas replied:
@kontsevoy can you address my comment here?

klizhentas replied:
k, trust you on this then :)

kontsevoy replied:
yeah, look at `NewTunClient()` implementation if you're curious. it checks its `addrStorage` and reads auth servers from there.

klizhentas commented on lib/srv/exec.go:
```diff
@@ -51,23 +57,38 @@ type execResponse struct {
 		ctx:     ctx,
 		cmdName: e.Command,
-		out:     &bytes.Buffer{},
+		isSCP:   isSCP,
```
is `isSCP` used anywhere?

klizhentas commented on lib/srv/sshserver.go:
```diff
@@ -692,6 +689,10 @@ func (s *Server) handleSessionRequests(sconn *ssh.ServerConn, ch ssh.Channel, in
 		case result := <-ctx.result:
 			ctx.Infof("ctx.result = %v", result)
+			// pass back stderr output
+			if result.stderr != nil {
```
probably more robust way to check to avoid zero writes would be `if len(result.stderr) != 0`

klizhentas commented on lib/sshutils/scp/scp.go:
```diff
@@ -339,29 +337,17 @@ func (cmd *Command) receiveDir(st *state, fc NewFileCmd, ch io.ReadWriter) error
-	return f == "scp"
-}
-
 func ParseCommand(arg string, conn ssh.ConnMetadata, alog events.IAuditLog) (*Command, error) {
```
looks like `conn` parameter is not used anymore then.

kontsevoy replied:
nice find: the entire function is not needed anymore

kontsevoy committed:
Direct execution of commands via SSH

Before this commit, command execution was done via "shell -c", now
Teleport executes them directly

kontsevoy committed:
Intermediate commit

kontsevoy committed:
Intermediate commit (scp up/down works agaisnt openSSH servers)

kontsevoy committed:
SCP error messages are proper now

`tsh scp` now reports the same error messages as OpenSSH scp:

- when talking to Teleport servers
- when talking to OpenSSH servers

kontsevoy committed:
Exec behaviour is more compatible with OpenSSH

klizhentas commented on lib/client/keyagent.go:
```diff
@@ -99,9 +99,13 @@ func (a *LocalKeyAgent) AddHostSignersToCache(hostSigners []services.CertAuthori
 func (a *LocalKeyAgent) CheckHostSignature(hostId string, remote net.Addr, key ssh.PublicKey) error {
+	// TODO (ev) remove this!
+	// we're temporarily turning off host validation to experiment with OpenSSH servers
+	//return nil
```
can you remove this commented out code?

klizhentas commented:
lgtm
